File: 1996/2/3-2.py
#!/usr/bin/env python

daysInMonth = (31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
days = ("Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday")

def jan1(mm, dd, yyyy):
    return ((36 + (yyyy - 1) + ((yyyy - 1) / 4) - ((yyyy - 1) / 100) + ((yyyy - 1) / 400)) % 7)

def leap(yyyy):
    return ((((yyyy % 4) == 0) and ((yyyy % 100) != 0)) or ((yyyy % 400) == 0))

def validDay(mm, dd, yyyy):
    return (daysInMonth[mm - 1] + (leap(yyyy) and mm == 2))

def valid(mm, dd, yyyy):
    return ((mm > 0) and (mm <= 12) and (dd > 0) and (dd <= validDay(mm, dd, yyyy)) and (yyyy > 0))

fin = file("prob2_in.txt", "r")
while 1:
    dp, dd, mm, yyyy = 0, 0, 0, 0
    line = fin.readline()
    if line == "":
        break
    if line[-1] == "\n":
        line = line[:-1]
    (mm, dd, yyyy) = line.split("/")
    mm = int(mm)
    dd = int(dd)
    yyyy = int(yyyy)
    if not valid(mm, dd, yyyy):
        print "Invalid date in the input."
        continue
    dp = sum(daysInMonth[0:mm-1]) + dd + jan1(mm, dd, yyyy) - 1
    print "%s/%s/%s falls on a %s" % (mm, dd, yyyy, days[dp % 7])

File: 1996/2/3-2.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

long daysInMonth[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
string days[] = {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};

int jan1(long mm, long dd, long yyyy)
{
    return ((36 + (yyyy - 1) + ((yyyy - 1) / 4) - ((yyyy - 1) / 100) + ((yyyy - 1) / 400)) % 7);
}

int leap(long yyyy)
{
    return (((((yyyy % 4) == 0) && ((yyyy % 100) != 0))) || ((yyyy % 400) == 0));
}

int validDay(long mm, long dd, long yyyy)
{
    return (daysInMonth[mm - 1] + (leap(yyyy) && mm == 2));
}

int valid(long mm, long dd, long yyyy)
{
    return ((mm > 0) && (mm <= 12) && (dd > 0) && (dd <= validDay(mm, dd, yyyy)) && (yyyy > 0));
}

int main()
{
    ifstream in("dates.dat");
    long mm, dd, yyyy, dp;
    dp = dd = mm = yyyy = 0;
    char ch;
    while (in >> mm >> ch >> dd >> ch >> yyyy)
    {
        if (!valid(mm,dd,yyyy))
        {
            cout << "Invalid date in the input." << endl;
            continue;
        }
        for (int i = 0; i < (mm - 1); i++)
            dp += daysInMonth[i];
        dp += dd;
        dp += jan1(mm,dd,yyyy);
        dp--;
        cout << mm << "/" << dd << "/" << yyyy << " falls on a " << days[dp % 7] << endl;
        dp = dd = mm = yyyy = 0;
    }
}
